# MotAli
Alignment of motifs to detect their similarity and co-occurrence

# Description

# Algorithm

# Input data
Two motifs, one Fasta file to recognize these two motifs, the threshold of the expected recognition rate for both motifs, tables "Threshols vs. Expected Recognition Rate" computed preliminary with the whole genome file of gene promoters. 
Each motif is either (a) the position frequency matrix required to compute position weight matrix (PWM model of motif), or (b) SiteGA model of motif, the list of locally positioned dinucleotides with their weights.

# Output data

# Source code and command line arguments

# Command line examples
